### DISCUSSION TITLE
Livewire  v3 - Livewire.dispatchTo not working

### DISCUSSION THREAD

**Discussion by spicygorilla** [livewire/livewire#6222] [1. Help]
Hi all,

I'm looking to replace the following on v2 with the requirements for v3.

```x-on:click="window.livewire.emitTo('record.details', 'showModal', '1')"```

I can see in the docs for V3: https://livewire.laravel.com/docs/javascript

```Livewire.dispatchTo('dashboard', 'post-created', { postId: 2 });```

This does not work for me at all.

```x-on:click="Livewire.dispatchTo('records.details', 'showModal', { record: 1})"```

I even tried using the window.Livewire.dispatchTo but I can't seem to get it working.

**The blade file listing records**

```
<dix x-data>
    <div>
        @foreach($records AS $record)
            <button 
                type="button" 
                x-on:click="Livewire.dispatchTo('records.details', 'openModal', {record: '{{ $record->id }}'})">
                    View details
            </button>
        @endforeach
    </div>
    <livewire:records.details />
</div>
```

The records.details Livewire component located at App/Livewire/Records/Details

```
namespace App\Livewire\Records;

use Illuminate\Contracts\View\View;
use Livewire\Component;

class Details extends Component
{
    public bool $showModal = false;
    public ?int $recordId;

    public function getListeners() : array
    {
        return ['openModal', 'closeModal'];
    }

    public function openModal($data) : void
    {
        $this->recordId = $data[0];

        $this->showModal = true;
    }

    public function closeModal() : void
    {
        $recordId = null;
        
        $this->showModal = false;
    }

    public function render() : View
    {
        return view('records.details');
    }
}
```

**The blade file for the details component** 

```
<div x-data="{ show: @entangle('showModal') }">
    <div x-show="show" x-cloak>
        <p>Modal is open</p>
        <button type="button" wire:click="closeModal">Close modal</button>
    </div>
</div>
```

V2 - everything works fine using livewire.emitTo, V3 - no dice. Any feedback on where I'm going wrong would be much appreciated. 

**Passing data to the component method being called**

Related to this is passing data to the method. In V2 using emitTo, we could pass a string, int, array etc. but it seems in v3, you always have to pass an  object? Example:

v2:

```
public function openModal(int $recordId) : void
{
    $this->recordId = $recordId;

    $this->showModal = true;
 }
```

v3:

```
public function openModal($data) : void
{
    // doing a dd on $data always give an array where the keys are removed.
    // I would expect recordId to be $data['record'],  but this does not exist, so I have to use $data[0]
    $this->recordId = $data[0];

    $this->showModal = true;
}
```

Is it possible to pass  a single property as string/int/float/bool to the method of a component we are dispatching to?

Thanks for your time!

**Comment by spicygorilla** [comment]
I managed to find a solution to one of my problems in a recent PR: https://github.com/livewire/livewire/pull/6198

```
    <button type="button" wire:click="$dispatchTo('demo-modal', 'openModal')">Open Modal window</button>

    <livewire:demo-modal />
```

I'm still trying to understand why the window.Livewire.dispatchTo isnt working in addition to

**Is it possible to pass a single property as string/int/float/bool to the method of a component we are dispatching to instead of an object?**

Thanks

**Reply by spicygorilla** [reply]
I'm still trying to understand why the window.Livewire.dispatchTo isnt working in addition to

This seems to work:

```
window.Livewire.dispatchTo('', 'records.details', 'showModal', { record: 1})"
```

It appears that dispatchTo takes 4 parameters:

```
component : empty string ''
componentName: 'records.details' 
name: 'showModal'
params: { record: 1}
```

If anyone can tell me what I am supposed to putting as the first parameter? Adding an empty string or using **this** seems to work ok.

**Comment by joshhanley** [comment]
@spicygorilla yeah there seems to be an issue with that. Any chance you could submit a failing test PR that uses the documentation example so we can investigate? Thanks!

**Reply by jacksontong** [reply]
https://github.com/livewire/livewire/blob/37f11583c61a75d51b2146c2fe38f506ad36014b/js/features/supportEvents.js#L67C28-L67C37
I don't see the first parameter is being used on the function if I pass an empty string it works
```js
window.Livewire.dispatchTo('', 'records.details', 'showModal', { record: 1})
```

**Reply by kaveeshHeshan** [reply]
This worked perfectly for me. Thank you @jacksontong

**Reply by danielreales7** [reply]
I have a question. I have this code:

`x-on:dblclick.prevent="window.Livewire.dispatchTo('', 'generateTicket', 'add-double-click')"`

My component is GenerateTicket. I don't know if I have to put generate-ticket or generateTicket. Third parameter I have this function:

`#[On('add-double-click')]
    public function addTicketWithDoubleClick()
    {
        $this->addToTicket();
    }`
    
Or I don't need to put #[On('add-double-click')] on function.

I've tried all the methods that have been mentioned, and it still doesn't work.

Thank you for your help!

**Comment by vladrusu** [comment]
Yeah, interesting, looking in Livewire's code, internal dispatchTo (not the $dispatchTo alias) takes 4 parameters, even tough the first one - component, is not used at all in the code. Weird...

```
  function dispatchTo(component, componentName, name, params) {
    let targets = componentsByName(componentName);
    targets.forEach((target) => {
      dispatchEvent(target.el, name, params, false);
    });
  }
```

Is this intended or is a bug? Maybe @calebporzio can help us.

**Comment by n0ah01** [comment]
I have a similar problem with $dispatchTo(), where the right function is called, but the arguments are not passed to the function:

`wire:click="$dispatchTo('kanban-detail', 'show_detail', { id: {{ $task_id }} })"`

```
#[On('show_detail')]
public function fill_detail( $asdf = null ) {
	dd( $asdf );
}
```
#![image](https://github.com/livewire/livewire/assets/50921507/a56ff9fc-2f53-4615-a472-1093f0f3cf27)

**Comment by vladrusu** [comment]
@n0ah01 Well, name your argument $id, not $asdf, or trigger dispatch to like so:
`wire:click="$dispatchTo('kanban-detail', 'show_detail', { asdf: {{ $task_id }} })"`

**Reply by n0ah01** [reply]
My bad, I thought the whole object is being passed as an argument...
Thanks alot!

**Reply by aquadancer** [reply]
wery thank you

**Comment by vladrusu** [comment]
BTW, I've added a pull request - #7099 - which addresses this issue.
Hopefully it gets merged.
